### ISSUE TITLE
Integration of complex.js

### ISSUE THREAD

**Issue by infusion** [josdejong/mathjs#560]
TL;DR: Check out the "complex" branch and see what we can do about the errors.

It was quite a lot of work, but after 500 additional lines in complex.js and > 1100 deleted lines in math.js, I have something to show and to discuss about. BTW: complex.js is still not more than 2kb when gzipped and minimized :)

Now to the bad news. I have still 61 unresolved tests, as I haven't touched them yet and that's something we should talk about. We can sort these issues into the following categories:

Number imprecision
---
This is the easiest thing, where we simply can change the tests on those numbers, where the last digit differs. :white_check_mark:

Number formatting
---
Complex.js doesn't have the feature of showing only a certain length of digits. If this feature is needed, I would suggest to overwrite the toString method in the Complex.js file in math.js. This would also fix the next issue.

Printed whitespaces
---
To stringify a number, Complex.js doesn't add whitespaces around the plus or minus operator. I added support to the parser, which now ignores whitespaces, but when it comes to outputting the numbers, I think it's cleaner to not add whitespaces. If this is really needed, then it's a formatting thing and a special use case, math.js should handle.

I mentioned it already with #550, I think parsing and outputting should be part of math.js at all. The build process could strip off the string parser from Fraction.js, Complex.js and the other type classes to save some bytes. Actually, you have this parser already, which decides what class should be used and so on.

from/toPolar function throws some exceptions
---
These two functions are the only two, which I haven't recycled. I left it in the Complex.js file of math.js, but it has some problems though.

Parser
---
Your string parser is a bit more robust. This means it is able to handle things like "3+-4i". As my library has this feature more or less as an little gimmick, I don't support these more complex inputs. What do you think, how these should be treated?

In order to comply with the math.js behavior, I removed an DIV/0 exception, with the result that a test case in my suite now gives "Infinity+Infinityi". What do you think of these?

Robert

**Comment by josdejong** [comment]
Looks promising @infusion ! I don't see any show stoppers, just small issues left.

## Number imprecision

yeah... round off errors. Many of the unit tests in math.js use `approx.equal` for these cases. It could be interesting to check *why* these results differ, as it's mostly some basic arithmetics and the old and new implementation should be more or less the same. Some functions have special cases [like sqrt](https://github.com/josdejong/mathjs/commit/2ef53ce0691e14b4f836032a4fc133ecddde700f) to prevent specific numerical issues.

## Number formatting

Overriding `toString` is a good option I think. Other solutions will just complicate things.

## Printed whitespaces

As for the spaces, I like them a lot, it gives better human readable output. Similar to having spaces after commas in lists like `[1, 2, 3, 4]`. It's just easier for the eye. 

## Parser

The complex parser in math.js is indeed quite robust and even understands weird inputs like "-2-+3i". It's acceptable when the parser would be a bit more restrictive, BUT it should give an error when it doesn't understand something, it should NEVER accept something illegal and silently give wrong output. I started out with a regexp too but in the end wrote a small parser, there are just so many tricky edge cases when using a regexp for something like this...

## Complex.format

I see you also stripped the function Complex.prototype.format, this is used though. I suppose this will stay in math.js and is not relevant for Complex.js

**Comment by infusion** [comment]
Glad you like it :)

In order to make it short: You're right about the whitespaces. I'll add a custom toString function to the PR. I also like your reasoning about the parser, I'll change this in complex.js. 

The imprecision comes from the trial of separating more edge cases and different ways to calculate things. To name just one function, the `exp()` function is one central point of the library for which I spend quite a lot of time on to make it better. In general, we should compare values if the old implementation gives "better" results.

**Comment by infusion** [comment]
Okay, the parser is done :)

**Comment by infusion** [comment]
I decided to change my `toString()`, even if I had to change the API this way. But you're right, it's prettier with spaces. We have 30 tests left :)

**Comment by infusion** [comment]
Do you have an idea, why toString isn't invoked here?

```
  9) unaryMinus should perform unary minus of a complex number:
     AssertionError: { re: -3, im: -2 } == '-3 - 2i'
      at Context.<anonymous> (test/function/arithmetic/unaryMinus.test.js:47:12)
```

**Comment by infusion** [comment]
**Zero** remaining issues :) I released Complex.js 1.9.6 and pushed the the last fixes to the complex branch. Unfortunately, it still fails on travis for two things, could you have a look there and merge it into dev? If everything is fine, give me a hint, I'll then publish the current state as 2.0.0

**Comment by josdejong** [comment]
Great news :) 

I checked out the failing tests travis-ci tests, there are two tests failing on node 0.1 and 0.12:

```
  1) cos should return the cosine of a complex number:
      AssertionError: false == true
      + expected - actual
      -false
      +true
      
      at Context.<anonymous> (test/function/trigonometry/cos.test.js:93:12)

  2) sin should return the sine of a complex number:
      AssertionError: { [String: '0'] re: 0, im: 0 } deepEqual { [String: '1e-50i'] re: 0, im: 1e-50 }
      + expected - actual
       {
      -  "im": 0
      +  "im": 1e-50
         "re": 0
       }
      
      at Context.<anonymous> (test/function/trigonometry/sin.test.js:85:12)
``` 

The two concerning unit tests check `sin` and `cos` for a very small value (where sin and cos should be approximately linear):

```js
assert.ok(cos(complex('1e-50+1e-50i')).im != 0);

assert.deepEqual(sin(complex('1e-50i')), complex(0, 1e-50));
```

I'm not sure why they are failing only on these old versions of node.js and if we can fix this numerical issue in Complex.js.

**Comment by josdejong** [comment]
@infusion I've fixed these two unit tests, I think `Math.sin` or `Math.sinh` had some inaccuracies on node v0.10 and 0.12.

I've merged `complex` into `develop` and removed the `complex` branch. Congrats :)

**Comment by infusion** [comment]
Do I have to fix something in Complex.js for the accuracy or is it just fine?

Otherwise, thanks for merging! :)

**Comment by josdejong** [comment]
No I think it's fine, I think it's the implementation of `Math.sinh` and relatives which seems to be inaccurate in old versions of node. The only way we could fix this is write our own implementations, which sounds silly because it's only needed for outdated versions of node.js and would be a step back for new versions of node.js.

**Comment by infusion** [comment]
Yep, I totally agree. Especially the node community is still in favor of always having the newest things. Finding old node versions in production should therefore be really seldom.

**Comment by infusion** [comment]
Okay, I published 2.0.0, you can change this in the package.json in order to force npm to upgrade the dependency.

**Comment by josdejong** [comment]
Sure I've updated. I don't understand your versioning though, why did you change the number to v2? Did you do something breaking since v1.9.7? I'm used to the semver versioning where a version number means `Breaking.Feature.Fix`

https://medium.com/javascript-scene/software-versions-are-broken-3d2dc0da0783#.2b711i3om
